ustack: add buildid field for client-side symbolization

alban opened the pull request:
# ustack: add buildid field for client-side symbolization

Executables in the ubuntu container image don't have symbol tables but they have a buildid that can be resolved on [debuginfod](https://sourceware.org/elfutils/Debuginfod.html) servers.

## How to use

```
sudo ig run ghcr.io/inspektor-gadget/gadget/trace_capabilities --collect-ustack --comm=chroot -o yaml

ustack:
  addresses: '[0]0x00007f46ba58e4fb; [1]0x00007f46ba4913b8; [2]0x00007f46ba49147b;
    [3]0x000055b3b6b0e2d5; '
  buildid: '[0]91f01b4ad171c80b6303d08d1f08cba8b990413d +1274fb; [1]91f01b4ad171c80b6303d08d1f08cba8b990413d
    +2a3b8; [2]91f01b4ad171c80b6303d08d1f08cba8b990413d +2a47b; [3]ae03c7096b4806aa173c0fd1861b019daf6a057a
    +32d5; '
  symbols: ""
```

Test workload:
```
docker run -ti --rm ubuntu:24.10 chroot /
```

The symbols can be resolved client-side with:

```
$ sudo ig run ghcr.io/inspektor-gadget/gadget/trace_capabilities \
	--collect-ustack --comm=chroot -o json | \
	jq --unbuffered -r .ustack.buildid | \
	./symbolizer.sh
__GI_chroot
__libc_start_call_main
call_init
_start
```

symbolizer.sh contains the following:
```bash
#!/bin/bash
export DEBUGINFOD_URLS="https://debuginfod.elfutils.org"
while IFS= read -r line; do
    stack="$(echo "${line%;}" | tr ';' '\n')"
    while IFS= read -r entry; do
        if [[ "$entry" =~ \[.*\]([0-9a-fA-F]+)\ \+(.*) ]]; then
            buildid="${BASH_REMATCH[1]}"
            offset="${BASH_REMATCH[2]}"
            sudo --preserve-env=DEBUGINFOD_URLS DEBUGINFOD_VERBOSE=false \
                debuginfod-find debuginfo $buildid &> /dev/null
            sudo addr2line  -e /root/.cache/debuginfod_client/$buildid/debuginfo -f -C $offset 2>/dev/null | head -1
        fi
    done <<< "$stack"
done
```

## Testing done


## TODO

- [x] Decide how to enable/disable the buildid field from the ebpf program
- [x] Client-side operator for symbolization with debuginfod.

mauriciovasquezbernal commented:
This looks amazing, now the thing is how to implement the client part. Have you envisioned how it should work? Do you think this operator will have the time to process the data online (i.e. resolving the symbols when they arrive) or should we enable a way to record all the information to a file and them enrich it offline?

alban commented:
> This looks amazing, now the thing is how to implement the client part. Have you envisioned how it should work? Do you think this operator will have the time to process the data online (i.e. resolving the symbols when they arrive) or should we enable a way to record all the information to a file and them enrich it offline?
> 
> Or what other ideas do you have to implement it client side?

I don't know... Maybe it is easier for Inspektor Gadget Desktop because it can download the debuginfo asynchronously and update the flamegraph when the data becomes available.

It is more difficult for ig/gadgetctl in a terminal because it can't update previous events it has already printed.

It might be ok for some gadgets like trace_mount to do it online because it does not generate a lot of events. But gadgets like cpu profilers should do the symbolization at least asynchronously or offline.

mauriciovasquezbernal reviewed (commented):
I did a first pass to get familiar with the code. I'll make a second one checking the UX.

alban commented on include/gadget/user_stack_map.h:
```diff
@@ -24,6 +24,23 @@ struct {
+	__uint(value_size,
+	       GADGET_USER_MAX_STACK_DEPTH * sizeof(struct bpf_stack_build_id));
+	__uint(max_entries, 0); // To be replaced at runtime
+} ig_build_id SEC(".maps");
```
Why? Other maps in include files are named ig_ustack and ig_kstack.

I see we also have `gadget_mntns_filter_map` and `gadget_sockets`...

I prefer shorter names because `bpftool map` will print truncated names.

alban commented on pkg/operators/ustack/ustack.go:
```diff
@@ -88,11 +109,17 @@ func (o *Operator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, in
+	if opts.UseSymtab || opts.UseDebugInfodClientCache {
+		// Use a sync.OnceValue to delay the creation of the Symbolizer because:
+		// - otherwise it is needlessly created during GetGadgetInfo
+		instance.symbolizer = sync.OnceValue(func() *symbolizer.Symbolizer {
```
Alternatively, we could implement some method on the gadgetContext to figure out if we're in a GetGadgetInfo context. Similar idea to #4816. That would enable earlier errors.

alban commented on pkg/operators/ustack/ustack.go:
```diff
@@ -173,13 +201,36 @@ func (o *OperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+			var addressesField, buildIDField, symbolsField datasource.FieldAccessor
+			var err error
+
+			if addressesFieldAll := in.GetSubFieldsWithTag("name:addresses"); len(addressesFieldAll) == 0 {
```
I can check for that with a second tag "operator:ustack" and warn & skip when the tag is not here.

alban commented on pkg/symbolizer/symbolizer.go:
```diff
@@ -187,109 +230,76 @@ type Task struct {
-	}
-	if fs.Size() > maxExecutableSize {
-		return nil, fmt.Errorf("executable is too large (%d bytes)", fs.Size())
+func (s *Symbolizer) resolveStackItemsWithTable(table *symbolTable, baseAddress uint64, stackItems []StackItemQuery, res []StackItemResponse) {
```
This method was added in a first commit and modified in a second commit. Do you know a tool to move it easily and keep the PR decomposed in logical commits? Otherwise, I'd need to run an interactive rebase, edit both commits, and resolve the conflicts.

alban replied:
I did it manually with these commands:
```
git rebase -i b710bb9f9200cf36222dbe978ca8e039c66816c4
# Add "edit" for the relevant commits

Iterate over each commit to edit:
    If conflict:
        git restore -s CURRENT_COMMIT_ID .

    sed -i 's/stackItems/stackQueries/g' $(git ls-files '*.go')

    If no conflict:
        git commit -a --amend --no-edit
    If conflict:
        git commit -a --no-edit

    git rebase --continue
```

alban commented on pkg/symbolizer/symbolizer.go:
```diff
@@ -339,3 +425,70 @@ func (s *Symbolizer) newSymbolTable(pid uint32, expectedExeKey exeKey) (*symbolT
+		s.symbolTablesFromBuildID[buildIDStr] = table
+
+		table.timestamp = time.Now()
+		symbol := table.lookupByAddr(stackItems[i].Offset)
```
I did it manually with these commands:
```
git rebase -i b710bb9f9200cf36222dbe978ca8e039c66816c4
# Add "edit" for the relevant commits

Iterate over each commit to edit:
    If conflict:
        git restore -s CURRENT_COMMIT_ID .

    sed -i 's/stackItems/stackQueries/g' $(git ls-files '*.go')

    If no conflict:
        git commit -a --amend --no-edit
    If conflict:
        git commit -a --no-edit

    git rebase --continue
```

alban commented:
I started preliminary smaller PRs:
- [ ] https://github.com/inspektor-gadget/inspektor-gadget/pull/4919
- [ ] https://github.com/inspektor-gadget/inspektor-gadget/pull/4920

alban commented on pkg/symbolizer/symbolizer.go:
```diff
@@ -339,3 +425,70 @@ func (s *Symbolizer) newSymbolTable(pid uint32, expectedExeKey exeKey) (*symbolT
+		symbol := table.lookupByAddr(stackQueries[i].Offset)
+		if symbol != "" {
+			res[i].Found = true
+			res[i].Symbol = symbol
```
If both symtab and debuginfod resolver worked, it probably does not matter which one we pick.

It does not usually happen: if the debuginfod method works, it means the binary comes from a Linux distribution which ships its binaries stripped and send the dbg pkg to the debuginfod database. There is no point to populate the debuginfod database if the binary is not stripped in the first place.

alban commented on pkg/operators/ustack/ustack.go:
```diff
@@ -270,16 +288,22 @@ func (o *OperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+							// The gadget didn't collect build ids
+							// Gadgets can use --collect-build-id to enable collecting build ids
+							break
+						}
```
To avoid one level of indentation :D 
Should I rewrite it to avoid the errLookup check in the loop?

mauriciovasquezbernal commented on pkg/operators/ustack/ustack.go:
```diff
@@ -20,8 +20,10 @@ import (
+	"unsafe"
 
 	"github.com/cilium/ebpf"
+	"golang.org/x/sys/unix"
```
This should to to ustack_linux.go

mauriciovasquezbernal commented on pkg/operators/ustack/ustack.go:
```diff
@@ -247,6 +254,46 @@ func (o *OperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+						if int(buildIDMapValueSize) != len(buildIDBuf) {
+							logger.Warnf("build id map has unexpected value size %d instead of %d", buildIDMapValueSize, len(buildIDBuf))
+							return nil
+						}
```
This is changed on the "ustack operator: add client-side support (gadgetctl, kubectl-gadget)" commit. Why not to squash "ustack: add buildid field for client-side symbolization", "symbolizer: move common functions to new file" & "ustack operator: add client-side support (gadgetctl, kubectl-gadget)"?

mauriciovasquezbernal commented on pkg/operators/ustack/ustack_linux.go:
```diff
@@ -47,6 +49,59 @@ func readUserStackMap(gadgetCtx operators.GadgetContext, userStackMap *ebpf.Map,
+		// struct bpf_stack_build_id is part of Linux UAPI:
+		// https://github.com/torvalds/linux/blob/v6.14/include/uapi/linux/bpf.h#L1451
+		type bpfStackBuildID struct {
+			status     int32
+			buildID    [unix.BPF_BUILD_ID_SIZE]uint8
+			offsetOrIP uint64 // Union of offset and ip
```
These fields need to be exported, otherwise it panicks:


```bash
panic: reflect: reflect.Value.SetInt using value obtained using unexported field

goroutine 21387 [running]:
reflect.flag.mustBeAssignableSlow(0xc0003bcd60?)
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/reflect/value.go:254 +0xb4
reflect.flag.mustBeAssignable(...)
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/reflect/value.go:244
reflect.Value.SetInt({0x3a422a0?, 0xc002b22000?, 0x4635080?}, 0x1)
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/reflect/value.go:2132 +0x45
encoding/binary.(*decoder).value(0xc0003bd238, {0x3a422a0?, 0xc002b22000?, 0x41?})
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/encoding/binary/binary.go:883 +0x13f
encoding/binary.(*decoder).value(0xc0003bd238, {0x3e2faa0?, 0xc002b22000?, 0xc0003bd1b0?})
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/encoding/binary/binary.go:863 +0x7e5
encoding/binary.(*decoder).value(0xc0003bd238, {0x3ace7e0?, 0xc002b22000?, 0xc0003bd268?})
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/encoding/binary/binary.go:850 +0x8cc
encoding/binary.Decode({0xc002b23000, 0xfe0, 0xfe0}, {0x45f8190, 0x63c0940}, {0x39ddb80, 0xc002b22000})
        /home/mauriciov/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/encoding/binary/binary.go:326 +0x345
github.com/cilium/ebpf/internal/sysenc.Unmarshal({0x39ddb80, 0xc002b22000}, {0xc002b23000, 0xfe0, 0xfe0})
        /home/mauriciov/go/pkg/mod/github.com/cilium/ebpf@v0.19.1-0.20250729164112-d994daa25101/internal/sysenc/marshal.go:91 +0x1cf
github.com/cilium/ebpf/internal/sysenc.Buffer.Unmarshal({0xc002b23000?, 0x39ddb80?}, {0x39ddb80?, 0xc002b22000?})
        /home/mauriciov/go/pkg/mod/github.com/cilium/ebpf@v0.19.1-0.20250729164112-d994daa25101/internal/sysenc/buffer.go:75 +0x4a
github.com/cilium/ebpf.(*Map).unmarshalValue(0xc0023648c0?, {0x39ddb80?, 0xc002b22000?}, {0xc002b23000?, 0x0?})
        /home/mauriciov/go/pkg/mod/github.com/cilium/ebpf@v0.19.1-0.20250729164112-d994daa25101/map.go:1631 +0x365
github.com/cilium/ebpf.(*Map).LookupWithFlags(0xc0023648c0, {0x3a422e0, 0xc002b00160}, {0x39ddb80, 0xc002b22000}, 0x0)
        /home/mauriciov/go/pkg/mod/github.com/cilium/ebpf@v0.19.1-0.20250729164112-d994daa25101/map.go:770 +0xe5
github.com/cilium/ebpf.(*Map).Lookup(...)
        /home/mauriciov/go/pkg/mod/github.com/cilium/ebpf@v0.19.1-0.20250729164112-d994daa25101/map.go:747
github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ustack.readUserStackMap({0x4616ce8?, 0xc000884000?}, 0xc002364870, 0xc0023648c0, 0x3317)
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/operators/ustack/ustack_linux.go:70 +0x458
github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ustack.(*OperatorInstance).init.func1({0x4629668?, 0xc0004cc9c0?}, {0x45cc020, 0xc002b12050})
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/operators/ustack/ustack.go:301 +0x795
github.com/inspektor-gadget/inspektor-gadget/pkg/datasource.(*dataSource).Subscribe.func1({0x0?, 0x0?}, {0x45cc048?, 0xc002b12050?})
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/datasource/data.go:595 +0x42
github.com/inspektor-gadget/inspektor-gadget/pkg/datasource.(*dataSource).EmitAndRelease(0xc0004cc9c0, {0x45cc048, 0xc002b12050})
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/datasource/data.go:664 +0xdf
github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf.(*Tracer).processEvent(0xc00201f170, {0x45ec280?, 0xc001af11d0?}, {0xc002c23c30, 0xa8, 0xa8})
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/operators/ebpf/tracer.go:209 +0x252
github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf.(*Tracer).receiveEvents(0xc00201f170, {0x4616ce8, 0xc000884000}, 0xc003567f98?)
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/operators/ebpf/tracer.go:156 +0x25c
created by github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf.(*ebpfInstance).runTracer in goroutine 1
        /home/mauriciov/kinvolk/ebpf/inspektor-gadget/pkg/operators/ebpf/tracer.go:260 +0x35f
```

mauriciovasquezbernal commented on pkg/operators/ustack/ustack_linux.go:
```diff
@@ -47,6 +49,59 @@ func readUserStackMap(gadgetCtx operators.GadgetContext, userStackMap *ebpf.Map,
+			offsetOrIP uint64 // Union of offset and ip
+		}
+		buildid := [ebpftypes.UserPerfMaxStackDepth]bpfStackBuildID{}
+		if buildIDMap.ValueSize() != uint32(unsafe.Sizeof(buildid)) {
+			logger.Warnf("build id map has unexpected value size %d instead of %d",
+				buildIDMap.ValueSize(), unsafe.Sizeof(buildid))
+			return "", "", nil, nil
+		}
```
I think we can move this check to the init phase to avoid having to perform it on the hot path.

mauriciovasquezbernal reviewed (commented):
@alban what do you think of https://github.com/inspektor-gadget/inspektor-gadget/pull/4314/files#r2344519335?

alban commented:
I updated the branch and addressed most of the comments, but I still need to address https://github.com/inspektor-gadget/inspektor-gadget/pull/4314/files#r2344519335 and test again.

I'll do that after #4921 is merged.

mauriciovasquezbernal reviewed (approved):
Some minor nits but LGTM already, great job @alban!
